tools/asset-sync/assets.Tests.ps1

```diff
@@ -0,0 +1,430 @@
+}
+
+AfterAll {
+  if(!($env:DISABLE_INTEGRATION_BRANCH_CLEANUP)){
+    DeInitialize-Integration-Branches
+  }
+}
+
+Describe "AssetsModuleTests" {
+  Context "Evaluate-Target-Dir" {
+    It "Should evaluate a root directory properly." {
+      $Value = Evaluate-Target-Dir -TargetPath (Join-Path $PSScriptRoot ".." "..")
+      $Value | Should -Be @($false, $false)
+    }
+    
+    It "Should evaluate a recording directory properly." {
+      $Value = Evaluate-Target-Dir -TargetPath $PSScriptRoot
+      $Value | Should -Be @($true, $false)
+    }
+  
+    It "Should evaluate an transitory directory properly." {
+      $Value = Evaluate-Target-Dir -TargetPath (Join-Path $PSScriptRoot "..")
+      $Value | Should -Be @($false, $false)
+    }
+
+    It "Should recognize root in a test directory."{
+      $files = @(
+        "sdk/assets.json"
+      )
+      $testLocation = Describe-TestFolder -AssetsJsonContent (Get-Basic-AssetsJson) -Files $files
+
+      $result = Evaluate-Target-Dir -TargetPath $testLocation
+      $result | Should -Be @($false, $true)
+
+      $result = Evaluate-Target-Dir -TargetPath (Join-Path $testLocation "sdk")
+      $result | Should -Be @($true, $false)
+    }
+  }
+
+  Context "Resolve-AssetsJson" {
+    It "Should find basic assets.json" {
+      $files = @(
+        "a/b.json",
+        "a/b/c.json"
+      )
+
+      $testLocation = Describe-TestFolder -AssetsJsonContent (Get-Basic-AssetsJson) -Files $files
+      
+      $testLocation.GetType().Name | Should -Be "String"
+
+      $Config = Resolve-AssetsJson -TargetPath $testLocation
+      $recordingLocation = $Config.AssetsJsonLocation
+      $recordingLocation | Should -Be (Join-Path $testLocation "assets.json")
+    }
+
+    It "Should should traverse upwards to find assets.json." {
+      $files = @(
+        "sdk/storage/",
+        "sdk/storage/assets.json",
+        "sdk/storage/azure-storage-blob/awesome.json"
+      )
+      $testLocation = Describe-TestFolder -AssetsJsonContent (Get-Basic-AssetsJson) -Files $files
+      
+      $Result = Resolve-AssetsJson -TargetPath (Join-Path $testLocation "sdk" "storage" "azure-storage-blob")
+      $recordingLocation = $Result.AssetsJsonLocation
+      $recordingLocation | Should -Be (Join-Path $testLocation "sdk" "storage" "assets.json")
+    }
+
+    It "Should be able to resolve the assets json based on CWD" {
+      $files = @(
+        "sdk/storage/",
+        "sdk/storage/assets.json",
+        "sdk/storage/azure-storage-blob/awesome.json"
+      )
+      $testLocation = Describe-TestFolder -AssetsJsonContent (Get-Basic-AssetsJson) -Files $files
+      
+      try {
+        Push-Location -Path (Join-Path $testLocation "sdk" "storage" "azure-storage-blob")
+        $Result = Resolve-AssetsJson
+        $recordingLocation = $Result.AssetsJsonLocation
+        $recordingLocation | Should -Be (Join-Path $testLocation "sdk" "storage" "assets.json")
+      }
+      finally {
+        Pop-Location
```

Comment:
Instead of having to add these `try/finally` blocks all over the tests, you could set a `$TEST_INVOKE_DIRECTORY` like constant in `BeforeAll` and add a `Set-Location` call to it in a `BeforeEach` block.

Reply:
This is a great suggestion. I will do something like this.

Reply:
Due to needing to customize within a test, I am not going to be able to go this direction. Recall that the test itself defines what it's testing. Maybe I define a function that I call from the test function, and that takes a script block as an input. That way I have a single function call.

```
    It "Should be able to resolve the assets json based on CWD" {
      $files = @(
        "sdk/storage/",
        "sdk/storage/assets.json",
        "sdk/storage/azure-storage-blob/awesome.json"
      )
      Run-Test -AssetsJsonContent (Get-Basic-AssetsJson) -Files $files {
          # Test Directory with $TEST_INVOKE_DIRECTORY now populated and navigated to
         
          # other script assertions.
      }
```

Reply:
Yeah I recall seeing that for some of the tests a before block would work, but other tests needed to do their own setup. So it might be complicating matters to do this operation in multiple places multiple times.